server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java

```diff
@@ -411,13 +417,34 @@ public KeywordFieldMapper build(MapperBuilderContext context) {
             }
             super.hasScript = script.get() != null;
             super.onScriptError = onScriptError.getValue();
+
+            var sourceKeepMode = this.sourceKeepMode.orElse(indexSourceKeepMode);
+            String offsetsFieldName;
+            if (context.isSourceSynthetic()
```

Comment:
Don't we need to check that `doc_values` are enabled?

Reply:
I really don't like how this check repeats a very similar one in `DocumentParser` that decides to apply ignored source.

Reply:
Good point, we need to check whether doc value is enabled. I updated this check and added tests.